File: packages/beecount_automation_android/android/src/main/kotlin/dev/beecount/extensions/automation_android/NotificationSourceRegistry.kt
package dev.beecount.extensions.automation_android

internal data class NativeNotificationSource(
    val id: String,
    val packageName: String,
    val displayName: String,
)

internal object NotificationSourceRegistry {
    val supported: List<NativeNotificationSource> = listOf(
        NativeNotificationSource("wechat", "com.tencent.mm", "微信"),
        NativeNotificationSource("alipay", "com.eg.android.AlipayGphone", "支付宝"),
        NativeNotificationSource("unionpay", "com.unionpay", "云闪付"),
        NativeNotificationSource("xiaomi_wallet", "com.miui.tsmclient", "小米钱包"),
        NativeNotificationSource("jd", "com.jingdong.app.mall", "京东"),
        NativeNotificationSource("meituan", "com.sankuai.meituan", "美团"),
        NativeNotificationSource("douyin", "com.ss.android.ugc.aweme", "抖音"),
        NativeNotificationSource("taobao", "com.taobao.taobao", "淘宝"),
    )

    private val byPackage = supported.associateBy { it.packageName }

    fun sourceFor(packageName: String): NativeNotificationSource? = byPackage[packageName]

    fun isSupported(packageName: String): Boolean = byPackage.containsKey(packageName)
}

File: packages/beecount_automation_android/lib/src/notification_source_registry.dart
final class NotificationSourceDescriptor {
  const NotificationSourceDescriptor({
    required this.id,
    required this.packageName,
    required this.displayName,
    required this.parserPhase,
  });

  final String id;
  final String packageName;
  final String displayName;
  final String parserPhase;
}

abstract final class NotificationSourceRegistry {
  static const List<NotificationSourceDescriptor> supported =
      <NotificationSourceDescriptor>[
    NotificationSourceDescriptor(
      id: 'wechat',
      packageName: 'com.tencent.mm',
      displayName: '微信',
      parserPhase: 'first',
    ),
    NotificationSourceDescriptor(
      id: 'alipay',
      packageName: 'com.eg.android.AlipayGphone',
      displayName: '支付宝',
      parserPhase: 'first',
    ),
    NotificationSourceDescriptor(
      id: 'unionpay',
      packageName: 'com.unionpay',
      displayName: '云闪付',
      parserPhase: 'second',
    ),
    NotificationSourceDescriptor(
      id: 'xiaomi_wallet',
      packageName: 'com.miui.tsmclient',
      displayName: '小米钱包',
      parserPhase: 'second',
    ),
    NotificationSourceDescriptor(
      id: 'jd',
      packageName: 'com.jingdong.app.mall',
      displayName: '京东',
      parserPhase: 'later',
    ),
    NotificationSourceDescriptor(
      id: 'meituan',
      packageName: 'com.sankuai.meituan',
      displayName: '美团',
      parserPhase: 'later',
    ),
    NotificationSourceDescriptor(
      id: 'douyin',
      packageName: 'com.ss.android.ugc.aweme',
      displayName: '抖音',
      parserPhase: 'later',
    ),
    NotificationSourceDescriptor(
      id: 'taobao',
      packageName: 'com.taobao.taobao',
      displayName: '淘宝',
      parserPhase: 'later',
    ),
  ];

  static NotificationSourceDescriptor? byPackageName(String packageName) {
    for (final source in supported) {
      if (source.packageName == packageName) return source;
    }
    return null;
  }
}
